### ISSUE TITLE
We should offer the same `[get/set]MultipleIntegers` as we offer for `[read/write]MultipleIntegers` methods on ByteBuffer

### ISSUE THREAD

**Issue by FranzBusch** [apple/swift-nio#2736]
Normally we provide all four variants of methods like `get/set/read/write` on the various `ByteBuffer` operations but we seem to missing the `get/set` variants for the `multipleInterger` based methods.

**Comment by Lukasa** [comment]
I think @weissi might make the case that this is a feature not a bug, I know he's been unhappy with the get/set methods for a while. I'm curious what he thinks.

**Comment by weissi** [comment]
> I think @weissi might make the case that this is a feature not a bug, I know he's been unhappy with the get/set methods for a while. I'm curious what he thinks.

I hate get/set's prominence, yes. 99.9% of their uses are either verbose or incorrect, usually both. Particularly

```
// correct: Something(buffer: buffer)
//
// incorrect and often seen in the wild
buffer.getSomething(at: 0, length: buffer.readableBytes)

// correct but verbose, seldomly seen
buffer.getSomething(at: buffer.readerIndex, length: buffer.readableBytes)
```

is both incorrect & verbose.

---

Now, there _are_ some niche use cases for our `get*/`set*`s but most people just use them because `read*/`write*` are mutating. So I think instead of adding `get*`/`set*`s eagerly we should offer `peek*` which doesn't take an `at: Int` parameter. So `peek*` ~= `read*` except that it doesn't read it off.

---

My thinking is, let's invest time in

- Better docs, every single `get*` should be discouraged in its docs
- `peek*` methods for cases where people need to take a part out without reading it out
- if we still have time left: add the missing `get*`/`set*`s with discouraging comments

**Comment by FranzBusch** [comment]
My use-case is the classic `peek` case where I need to look into an incoming packet to decide the kind of it and the send into the right child channel.

**Comment by weissi** [comment]
> My use-case is the classic `peek` case where I need to look into an incoming packet to decide the kind of it and the send into the right child channel.

Makes sense, I think we should add `peek*`s for everything. That's btw also a good starter task because it really just is

```
@inlinable
public func peekX(length: Int) -> X? {
    return self.getX(at: self.readerIndex, length: length)
}
```

or

```
@inlinable
public func peekX(length: Int) -> X? {
    var `self` = self
    return self.readX(length: length)
}
```

**Comment by FranzBusch** [comment]
Yeah I am totally fine rebranding the current `get` methods to peek and avoid doing public `get/set` all together.

**Comment by glbrntt** [comment]
I really like the idea of `peek`, I think it covers most of the existing uses of `get`.

I worry slightly that `get` is still the obvious name to reach for and many users won't actually read the docs which should discouraging them from using it. Ideally I think `get`/`set` would still exist but on a view type so they're less prominent in the API.

**Comment by natikgadzhi** [comment]
Would it be too big of a change to add `peekX` and soft-deprecate get/set friends?

**Comment by Lukasa** [comment]
Offering peek is a good step. I’m not sure about deprecating get/set though: they’re genuinely useful.

**Comment by weissi** [comment]
> Offering peek is a good step. I’m not sure about deprecating get/set though: they’re genuinely useful.

Agreed! Maybe in the future (say a NIO3 release), we could give get/set a less prominent space in the API, as in: more verbiage to discourage accidental usage.

**Comment by natikgadzhi** [comment]
Ok! I’ll write up two issues then:

1. Provide peek* for all types that currently have corresponding get*
2. Change verbiage in docs around get/set to explain why using them in most cases is a bad idea, guide users to read* instead. I think I have a snippet from @weissi in my first PR a year ago.

**Comment by RaghavRoy145** [comment]
Hello! I have been playing around with this, and I added an implementation for Integers, I'm not sure if the scope of this is "every type that supports get* should get a peek*", and further, if this sort of implementation should be captured within a single PR. But I'd love to know if I can expand on the current implementation in that way if this is still something that is required (and if my implementation even makes sense!).

Thanks!

**Comment by glbrntt** [comment]
> I'm not sure if the scope of this is "every type that supports get* should get a peek*", 

That's the scope: there should be a peek for every get.

> and further, if this sort of implementation should be captured within a single PR. 

Now that you've done integers I think it's fine for this two be at least two separate PRs:
1. For the multiple-integer code which the issue was originally about, because that code is generated from a script (https://github.com/apple/swift-nio/blob/main/dev/generate-bytebuffer-multi-int.sh)
2. All other get methods on `ByteBuffer`

It's no problem if you want to split it up into more though!

**Comment by RaghavRoy145** [comment]
> > I'm not sure if the scope of this is "every type that supports get* should get a peek*",
> 
> That's the scope: there should be a peek for every get.
> 
> > and further, if this sort of implementation should be captured within a single PR.
> 
> Now that you've done integers I think it's fine for this two be at least two separate PRs:
> 
>     1. For the multiple-integer code which the issue was originally about, because that code is generated from a script (https://github.com/apple/swift-nio/blob/main/dev/generate-bytebuffer-multi-int.sh)
> 
>     2. All other get methods on `ByteBuffer`
> 
> 
> It's no problem if you want to split it up into more though!

Thanks for the clarification !

I think since it's more involved for multiple-integer (doesn't have a get* to work with trivially?), I can start with the other types with get methods so I don't need to worry about read indices being moved. Since I will also be writing tests for them, I wanted to know if the tests I have written for https://github.com/apple/swift-nio/pull/3157 are interesting enough. I find them alright to get started with, but I'd love to know if there's some cases that I should take care of that I may have missed?

Otherwise, I can create another PR that takes care of "All other get methods on `ByteBuffer`" to start with.
